### DISCUSSION TITLE
How to gracefully configure the environment in pyproject.toml?

### DISCUSSION THREAD

**Discussion by JulyFinal** [astral-sh/rye#1296] [Q&A]
Previously, I manually imported the contents of these folders (the current directory and the bt directory, etc.) by using 

`PYTHONPATH=$(pwd):$(pwd)/bt:$(pwd)/bt/b_service:$(pwd)/bt/t_service`

This is because my project has many sub-projects, and I want to add them to the `PYTHONPATH` in bulk.

so i try :
```
[tool.rye.workspace]
members = [
  "a",
  "b",
  "bt/*",
]
```

or
```
[tool.hatch.build.targets.wheel]
packages = [
  "a",
  "b",
  "bt/*",
]
```

But this doesn't seem to work. How should I set it up to correctly run rye run "my-rye-script" ?

and now i get this error msg 

```
Traceback (most recent call last):
  File "./my-project/bt/s/s.py", line 9, in <module>
    from a.xx import yy
ModuleNotFoundError: No module named 'a'
```

I'm still a newbie in this area, so any advice would be appreciated.

**Comment by bluss** [comment]
The workspace solution should work - if you have a "rye init"ed pyproject for each of the directories a, b, bt/b_service, bt/t_service and their requisite folder structure (the usual one is project/src/package_name, as you can see rye init creates).

Workspace members is not for python *packages* (those directories with `__init__.py`) it is for *projects* (those with pyproject.toml).


As for the hatch solution - that should work too - what's the error there? That's without creating a workspace and multiple projects. More information needed, maybe a minimal reproducing example.

**Reply by JulyFinal** [reply]
First. thanks for your help.

In fact, my project consists of two parts.

Project 1(project1): Contains two subfolders, a and b.

Project 2(bt): Contains two subfolders, b_service and t_service.

A Python script in the b folder imports some methods from the a folder.

Right now, I clone Project 2 into Project 1 first, then use the above method to make the script import the methods. 
`PYTHONPATH=$(pwd):$(pwd)/bt:$(pwd)/bt/b_service:$(pwd)/bt/t_service`

This my project dir tree.
```bash
project1
├── README
├── pyproject.toml
├── a
│   ├── __init__.py
│   └── ...
├── b
│   ├── __init__.py
│   └── ...
├── ...
└── bt(project2)
    ├── pyproject.toml
    ├── README
    ├── b_service
    │   ├── __init__.py
    │   └── my-rye-script.py
    └── t_service
        ├── __init__.py
        └── my-rye-script.py

```

my-rye-script.py
```python
from a.xx import yy
...
```


Maybe I shouldn't call these subprojects. 
My project doesn't have a structure like `src` etc. , and unfortunately, I can't control that.

**Reply by bluss** [reply]
That should be a use case for workspace. In the root (project1) set it up as a workspace, with members = ["bt"]. Then rye sync should set up the virtualenv correctly. (Assuming the pyproject.tomls are already setup right to install their respective packages).

Is `my-rye-script.py` installed as a script? Should be possible to run it with rye run <the script endpoint name>

**Reply by JulyFinal** [reply]
Thanks, it successfully ran. 😄

**Reply by Veronicandemo** [reply]
I'm getting the same but I'm only working under one project. In my case I have created my rye project as a package. The error i'm getting is when I try using the package inside a service . In folder structure I have the src which has two sub folders. When I import from the sub folders to another. I get  `ModuleNotFoundError: No module named 'src'`. This is if say I have a folder xy and another yz. If i try using imports from xy in yz i get the above error. The [tool.rye.workspace] solution is not working for me.

**Reply by bluss** [reply]
@Veronicandemo more details needed to give further guidance, but with two subfolders in src, that's two separate python packages (which is fine), and their names are given by the folder names. The regular way to import is to use "import xy" or "import yz". `src` does never occur in an import line, i.e it's never "import src" or "import src.xy".
